pachd crash upon startup

From a user running 1.4.3:

```
2017-04-11T23:50:47Z ERROR error discovering workers: watcher for prefix pachyderm_pps/workers/pipeline-pachsupport-boxes-flatten-v1/ closed for unknown reasons; retrying in 387.742923ms
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0x1695d12]
goroutine 110 [running]:
github.com/pachyderm/pachyderm/src/server/pps/server.(*apiServer).pipelineWatcher.func1(0xc42043bec0, 0xc4201f0748)
        /go/src/github.com/pachyderm/pachyderm/src/server/pps/server/api_server.go:859 +0x1c2
github.com/pachyderm/pachyderm/src/server/pkg/backoff.RetryNotify(0xc420264b40, 0x25dc760, 0xc42043bec0, 0xc420d71f98, 0xc4201f07a8, 0xa3eb25)
        /go/src/github.com/pachyderm/pachyderm/src/server/pkg/backoff/retry.go:35 +0x3f
github.com/pachyderm/pachyderm/src/server/pps/server.(*apiServer).pipelineWatcher(0xc420497540, 0x7f2148cdd360, 0xc420390f80, 0x6)
        /go/src/github.com/pachyderm/pachyderm/src/server/pps/server/api_server.go:897 +0xed
created by github.com/pachyderm/pachyderm/src/server/pps/server.(*apiServer).AddShard
        /go/src/github.com/pachyderm/pachyderm/src/server/pps/server/api_server.go:1573 +0x30f
```

Might be related:

```
 in 311.310947ms
2017-04-12T00:34:42Z INFO  goro for worker 10.244.4.59 is exiting
2017-04-12T00:34:42Z INFO  goro for worker 10.244.4.60 is exiting
2017-04-12T00:34:42Z INFO  goro for worker 10.244.2.42 is exiting
2017-04-12T00:34:42Z INFO  goro for worker 10.244.0.45 is exiting
2017-04-12T00:34:42Z ERROR error discovering workers: watcher for prefix pachyderm_pps/workers/pipeline-pachsupport-boxes-flatten-v1/ closed for unknown reasons; retrying in 627.662717ms
2017-04-12T00:34:42Z INFO  goro for worker 10.244.0.11 is exiting
2017-04-12T00:34:42Z INFO  goro for worker 10.244.2.8 is exiting
2017-04-12T00:34:42Z INFO  goro for worker 10.244.3.10 is exiting
2017-04-12T00:34:42Z INFO  goro for worker 10.244.3.9 is exiting
2017-04-12T00:34:42Z INFO  goro for worker 10.244.4.13 is exiting
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0x1696962]
goroutine 113 [running]:
github.com/pachyderm/pachyderm/src/server/pps/server.(*apiServer).jobWatcher.func1(0xc420278a20, 0xc420511748)
        /go/src/github.com/pachyderm/pachyderm/src/server/pps/server/api_server.go:931 +0x1c2
github.com/pachyderm/pachyderm/src/server/pkg/backoff.RetryNotify(0xc4202b1cb0, 0x25dc760, 0xc420278a20, 0xc421a07f98, 0xc4205117a8, 0xa3ead7)
        /go/src/github.com/pachyderm/pachyderm/src/server/pkg/backoff/retry.go:35 +0x3f
github.com/pachyderm/pachyderm/src/server/pps/server.(*apiServer).jobWatcher(0xc4200bdb80, 0x7f6d811a2a40, 0xc4204ca380, 0x6)
        /go/src/github.com/pachyderm/pachyderm/src/server/pps/server/api_server.go:970 +0xed
created by github.com/pachyderm/pachyderm/src/server/pps/server.(*apiServer).AddShard
        /go/src/github.com/pachyderm/pachyderm/src/server/pps/server/api_server.go:1592 +0x2c9
```

More logs:

```
panic: send on closed channel
goroutine 270 [running]:
github.com/pachyderm/pachyderm/src/server/pkg/collection.(*readonlyCollection).WatchByIndex.func1.1(0xc420360360, 0xc420627fd0, 0x25dc8e0, 0xc42045f610)
        /go/src/github.com/pachyderm/pachyderm/src/server/pkg/collection/collection.go:392 +0xb3
github.com/pachyderm/pachyderm/src/server/pkg/collection.(*readonlyCollection).WatchByIndex.func1(0xc420360360, 0x25dc8e0, 0xc42045f610, 0xc4203603c0, 0xc4205f0c00, 0x25d2660, 0xc42000e3c0)
        /go/src/github.com/pachyderm/pachyderm/src/server/pkg/collection/collection.go:419 +0x5bc
created by github.com/pachyderm/pachyderm/src/server/pkg/collection.(*readonlyCollection).WatchByIndex
        /go/src/github.com/pachyderm/pachyderm/src/server/pkg/collection/collection.go:439 +0x1b3
```

--- a/Makefile
+++ b/Makefile
@@ -148,10 +148,10 @@ launch-dev-bench: docker-build docker-build-test install
 
 push-bench-images: install
 	# We need the pachyderm_compile image to be up to date
-	docker tag pachyderm_pachd pachyderm/pachd:`pachctl version 2>/dev/null | head -n 2 | tail -n 1 | awk -v N=2 '{print $$N}'`-`git log | head -n 1 | cut -f 2 -d " "`
-	docker push pachyderm/pachd:`pachctl version 2>/dev/null | head -n 2 | tail -n 1 | awk -v N=2 '{print $$N}'`-`git log | head -n 1 | cut -f 2 -d " "`
-	docker tag pachyderm_worker pachyderm/worker:`pachctl version 2>/dev/null | head -n 2 | tail -n 1 | awk -v N=2 '{print $$N}'`-`git log | head -n 1 | cut -f 2 -d " "`
-	docker push pachyderm/worker:`pachctl version 2>/dev/null | head -n 2 | tail -n 1 | awk -v N=2 '{print $$N}'`-`git log | head -n 1 | cut -f 2 -d " "`
+	docker tag pachyderm_pachd pachyderm/pachd:`pachctl version 2>/dev/null | head -n 2 | tail -n 1 | awk -v N=2 '{print $$N}'`
+	docker push pachyderm/pachd:`pachctl version 2>/dev/null | head -n 2 | tail -n 1 | awk -v N=2 '{print $$N}'`
+	docker tag pachyderm_worker pachyderm/worker:`pachctl version 2>/dev/null | head -n 2 | tail -n 1 | awk -v N=2 '{print $$N}'`
+	docker push pachyderm/worker:`pachctl version 2>/dev/null | head -n 2 | tail -n 1 | awk -v N=2 '{print $$N}'`
 	docker tag pachyderm_test pachyderm/bench:`git log | head -n 1 | cut -f 2 -d " "`
 	docker push pachyderm/bench:`git log | head -n 1 | cut -f 2 -d " "`
 	

--- a/src/server/pkg/collection/collection.go
+++ b/src/server/pkg/collection/collection.go
@@ -384,15 +384,14 @@ func (c *readonlyCollection) WatchByIndex(index Index, val interface{}) (watch.W
 	}
 	go func() (retErr error) {
 		defer func() {
-			close(eventCh)
 			if retErr != nil {
 				eventCh <- &watch.Event{
 					Type: watch.EventError,
 					Err:  retErr,
 				}
-				close(eventCh)
 				watcher.Close()
 			}
+			close(eventCh)
 		}()
 		for {
 			var ev *watch.Event

--- a/src/server/pps/server/api_server.go
+++ b/src/server/pps/server/api_server.go
@@ -871,7 +871,10 @@ func (a *apiServer) pipelineWatcher(ctx context.Context, shard uint64) {
 			return err
 		}
 		for {
-			event := <-pipelineWatcher.Watch()
+			event, ok := <-pipelineWatcher.Watch()
+			if !ok {
+				return fmt.Errorf("pipelineWatcher closed unexpectedly")
+			}
 			if event.Err != nil {
 				return event.Err
 			}
@@ -927,7 +930,10 @@ func (a *apiServer) jobWatcher(ctx context.Context, shard uint64) {
 		}
 
 		for {
-			event := <-jobWatcher.Watch()
+			event, ok := <-jobWatcher.Watch()
+			if !ok {
+				return fmt.Errorf("jobWatcher closed unexpectedly")
+			}
 			if event.Err != nil {
 				return event.Err
 			}

--- a/src/server/pps/server/api_server.go
+++ b/src/server/pps/server/api_server.go
@@ -870,6 +870,7 @@ func (a *apiServer) pipelineWatcher(ctx context.Context, shard uint64) {
 		if err != nil {
 			return err
 		}
+		defer pipelineWatcher.Close()
 		for {
 			event, ok := <-pipelineWatcher.Watch()
 			if !ok {
@@ -928,7 +929,7 @@ func (a *apiServer) jobWatcher(ctx context.Context, shard uint64) {
 		if err != nil {
 			return err
 		}
-
+		defer jobWatcher.Close()
 		for {
 			event, ok := <-jobWatcher.Watch()
 			if !ok {

--- a/src/server/pps/server/api_server.go
+++ b/src/server/pps/server/api_server.go
@@ -1110,7 +1110,7 @@ func (a *apiServer) pipelineManager(ctx context.Context, pipelineInfo *pps.Pipel
 			return err
 		default:
 		}
-		protolion.Errorf("error running pipelineManager: %v; retrying in %v", err, d)
+		protolion.Errorf("error running pipelineManager for pipeline %s: %v; retrying in %v", pipelineInfo.Pipeline.Name, err, d)
 		if err := a.updatePipelineState(ctx, pipelineName, pps.PipelineState_PIPELINE_RESTARTING); err != nil {
 			protolion.Errorf("error updating pipeline state: %v", err)
 		}
@@ -1493,7 +1493,7 @@ func (a *apiServer) jobManager(ctx context.Context, jobInfo *pps.JobInfo) {
 		default:
 		}
 
-		protolion.Errorf("error running jobManager: %v; retrying in %v", err, d)
+		protolion.Errorf("error running jobManager for job %s: %v; retrying in %v", jobInfo.Job.ID, err, d)
 
 		// Increment the job's restart count
 		_, err = col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {

